eng/common/pipelines/templates/jobs/generate-job-matrix.yml

```diff
@@ -0,0 +1,139 @@
+parameters:
+- name: AdditionalParameters
+  type: object
+- name: DependsOn
+  type: object
+  default: null
+- name: CloudConfig
+  type: object
+  default: {}
+- name: MatrixConfigs
+  type: object
+  default: []
+- name: MatrixFilters
+  type: object
+  default: []
+- name: MatrixReplace
+  type: object
+  default: {}
+- name: JobTemplatePath
+  type: string
+# Set this to false to do a full checkout for private repositories with the azure pipelines service connection
+- name: SparseCheckout
+  type: boolean
+  default: true
+- name: SparseCheckoutPaths
+  type: object
+  default: []
+- name: Pool
+  type: string
+  default: azsdk-pool-mms-ubuntu-2204-general
+- name: OsVmImage
+  type: string
+  default: ubuntu-22.04
+- name: Os
+  type: string
+  default: linux
+# This parameter is only necessary if there are multiple invocations of this template within the SAME STAGE.
+# When that occurs, provide a name other than the default value.
+- name: GenerateJobName
+  type: string
+  default: 'generate_job_matrix'
+- name: PreGenerationSteps
+  type: stepList
+  default: []
+# Mappings to OS name required at template compile time by 1es pipeline templates
+- name: Pools
+  type: object
+  default:
+    - name: LinuxPool
+      os: linux
+    - name: LinuxNextPool
+      os: linux
+    - name: WindowsPool
+      os: windows
+    - name: MacPool
+      os: macOS
+
+jobs:
+- job: ${{ parameters.GenerateJobName }}
+  variables:
+    - template: /eng/pipelines/templates/variables/image.yml
```

Comment:
`image.yml` should be part of this PR I think.

Reply:
Originally the idea was to have image.yml be repo specific, but it is a hard requirement since this has a reference.

Reply:
@weshaggard maybe this is something we do want in eng/common just to avoid hidden file dependencies in our templates. However if we wanted to roll out image changes per-repo it would not work.

Reply:
I don't mind having a default one in eng\common and perhaps we can come up with a way to override the image in the language repo if needed. We could even make this a default template parameter that could be set to something as needed.

Reply:
One way would be to have the underlying values get set via coalesce e.g. `$[ coalesce(variables.LinuxPoolOverride, variables.LinuxPool) ]` but I'm not too excited about the config variable bloat of doubling all the values with `*Override`.